From 20615d597da69a5cf1f776c65c1971a6954185b2 Mon Sep 17 00:00:00 2001
From: Fabien Dessenne <fabien.dessenne@st.com>
Date: Wed, 16 Dec 2015 09:51:35 +0100
Subject: [PATCH 1/4] modetest: consider supported formats before selecting a
 DRM plane

This fixes an issue where the DRM planes do not support the same pixel
formats.
The current implementation selects a DRM plane without checking whether
the pixel format is supported or not. As a consequence modetest may try
to set up a plane not supporting the user request-format, which fails.
Modetest has to check the supported formats accross the plane list before
selecting a candidate.

Signed-off-by: Fabien Dessenne <fabien.dessenne@st.com>

Change-Id: If1bcbf9b7261bb3c86b560f79b57760b11e087f8
Reviewed-on: https://gerrit.st.com/46617
Tested-by: Fabien DESSENNE <fabien.dessenne@st.com>
Reviewed-by: Christophe PRIOUZEAU <christophe.priouzeau@st.com>
Reviewed-by: Fabien DESSENNE <fabien.dessenne@st.com>
---
 tests/modetest/modetest.c | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/tests/modetest/modetest.c b/tests/modetest/modetest.c
index 21d5438..5539253 100644
--- a/tests/modetest/modetest.c
+++ b/tests/modetest/modetest.c
@@ -963,7 +963,7 @@ static int set_plane(struct device *dev, struct plane_arg *p)
 	int crtc_x, crtc_y, crtc_w, crtc_h;
 	struct crtc *crtc = NULL;
 	unsigned int pipe;
-	unsigned int i;
+	unsigned int i, j;
 
 	/* Find an unused plane which can be connected to our CRTC. Find the
 	 * CRTC index first, then iterate over available planes.
@@ -986,8 +986,11 @@ static int set_plane(struct device *dev, struct plane_arg *p)
 		if (!ovr || !format_support(ovr, p->fourcc))
 			continue;
 
-		if ((ovr->possible_crtcs & (1 << pipe)) && !ovr->crtc_id)
-			plane_id = ovr->plane_id;
+		if ((ovr->possible_crtcs & (1 << pipe)) && !ovr->crtc_id) {
+			for (j = 0; j < ovr->count_formats; j++)
+				if (!strncmp(p->format_str, (char *) &ovr->formats[j], 4))
+					plane_id = ovr->plane_id;
+		}
 	}
 
 	if (!plane_id) {
-- 
1.9.1

